cc_grub_dpkg: determine idevs in more robust manner with grub-mkdevicemap

matthewruffell opened the pull request:
Please read https://bugs.launchpad.net/cloud-init/+bug/1877491

Replace the hardcoded list of devices with a more robust way of determining
the device which grub is installed to.

We use grub-probe to fetch the underlying disk the /boot directory is located on, and attempt to match the disk with its /dev/disk/by-id value. If no such /dev/disk/by-id/ value exists, we fallback to the plain disk name.

The changes are robust to unstable kernel device names and ordering, and use
/dev/disk/by-id values to populate grub-pc/install_devices where possible.

LP: #1877491

OddBloke commented:
Thanks for the submission, Matthew! There's some discussion of this approach happening in the bug, so I'm marking this "wip" until we've reached a consensus.

matthewruffell commented:
This new revision drops nearly all the complexity from the previous version.

Changes:
- no longer use grub-mkdevicemap as per xnox's recommendations due to it being deprecated
- the fork between BIOS and EFI is completely removed, and all instances use the same single code path
- we use grub-probe to determine the disk where the /boot directory is located, since it works for both BIOS and EFI systems
- removed duplicate code for the container usecase
- closely follow the algorithm from usable_partitions(), device_to_id() and available_ids() functions in the grub2 postinst.in file: https://paste.ubuntu.com/p/vKFNSwNyhP/

OddBloke commented:
Thanks for the review @xnox!

Just for clarity, @matthewruffell, I asked Dimitri to review the PR from a grub POV. I'm going to follow-up with a code review from a cloud-init perspective in the next day or two. (We have no indication that this is a high priority, so I've just added this to my TODO list for now. If there's some time sensitivity here that I'm not aware of, please let me know!)

matthewruffell commented:
Thank you for the review @xnox, and for your insight into udevadm as an alternative to walking symlinks.

@OddBloke Thanks for being clear with next steps. There is some time sensitivity, but nothing major, with SLA expiry coming up mid next month on 2020-06-14. The bug this patch fixes is more of a nuisance than anything, as it ensures that user's don't see the interactive dpkg prompt for selecting an install location when they upgrade their grub2 packages.

I will be available for quick iterations if you would like anything changed, just let me know.

As for getting this SRU'd into Ubuntu releases, do I need to submit any patches, or do you take care of that? I read https://wiki.ubuntu.com/CloudinitUpdates and it seems you backport the current tree. I'm happy to help test regardless.

OddBloke commented:
> @OddBloke Thanks for being clear with next steps. There is some time sensitivity, but nothing major, with SLA expiry coming up mid next month on 2020-06-14. The bug this patch fixes is more of a nuisance than anything, as it ensures that user's don't see the interactive dpkg prompt for selecting an install location when they upgrade their grub2 packages.
> 
> I will be available for quick iterations if you would like anything changed, just let me know.

I plan on reviewing this after lunch, so I'll let you know, thanks!

> As for getting this SRU'd into Ubuntu releases, do I need to submit any patches, or do you take care of that? I read https://wiki.ubuntu.com/CloudinitUpdates and it seems you backport the current tree. I'm happy to help test regardless.

Yep, this would go back as part of a regular SRU. That said, we're about to start an SRU process and we wouldn't expect to do another one before the described deadline. I'm wondering, @blackboxsw, if we should hold on the SRU until this lands. @matthewruffell, what are the target Ubuntu releases that the SLA applies to?

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         status_cb=False, capture=True)[0].strip()
```
The default value of `status_cb` is `None`, so I don't know that we need to pass `False` in here.

matthewruffell replied:
I actually thought `status_cb` did something completely different, but upon reading the implementation of uitl.subp() I was wrong. Latest commit drops `status_cb` completely.

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
+                         status_cb=False, capture=True)[0].strip()
+        # if /dev/disk/by-id/ exists, attempt to determine robust name
+        devlist = ""
+        with suppress(FileNotFoundError):
```
What `FileNotFoundError`s are we suppressing here? I'd prefer for us to scope this more narrowly, so it's clear that we won't be suppressing errors we don't mean to.

matthewruffell replied:
We suppress `FileNotFoundError` since `os.listdir()` throws `FileNotFoundError` if the directory does not exist. We cannot assume that `/dev/disk/by-id` will exist on a system, since most KVM based instances do not have this directory at all. 

I really wanted to avoid suppressing this error too, but it cannot be done.
I wish `os.listdir()` just returned None or a empty iterator instead, but since it throws the error, we need to suppress it.

OddBloke replied:
Aha, in that case, what do you think about specifically checking for its existence instead:

```suggestion
        by_id_path = "/dev/disk/by-id"
        if os.path.exists(by_id_path):
```

matthewruffell replied:
With the change to udevadm we no longer need to walk symlinks.

matthewruffell commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
+        devlist = util.subp(['awk', '{print $1}'],
+                            data=devlist, capture=True)[0].strip()
+        for device in devlist.split():
+            if str(Path(device).resolve()) == disk:
```
This has been resolved in the latest commit. We now resolve the symlink just once, and turn the full path to a string.

matthewruffell commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
+                devlist += "{} {}\n".format(device, Path(device).resolve())
+        devlist = util.subp(['sort', '-k2', '-s', '-u'],
+                            data=devlist, capture=True)[0].strip()
+        devlist = util.subp(['awk', '{print $1}'],
```
`devlist` is now a Python list of tuples. Each tuple is a (device, resolved symlink) pair. An example is: ('/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e', '/dev/nvme1n1')
The subprocess calls to sort and awk have now been replaced with a pure Python implementation.

matthewruffell commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +65,37 @@ def handle(name, cfg, _cloud, log, _args):
+            for device in sorted(os.listdir("/dev/disk/by-id/")):
+                device = "{}{}".format("/dev/disk/by-id/", device)
+                devlist += "{} {}\n".format(device, Path(device).resolve())
+        devlist = util.subp(['sort', '-k2', '-s', '-u'],
```
In the old, outdated commit this code references, `os.listdir()` does not make any attempt to returned a sorted list of files in the directory, so you must sort them yourself. This was important, as sort -k2 -s -u would use the sorted list to select the item that gets returned from the -u (unique) checks, and culls the rest.

In terms of the new commit, there are 2.5 sorts performed.

The first is `devlist.sort(key=lambda tup: tup[1])`. This sorts on the second tuple, which is the device. This groups together all the devices in the system, and we get output like:

```
[('/dev/disk/by-id/nvme-nvme.1d0f-4157533234443438343530393135464141303136-416d617a6f6e20454332204e564d6520496e7374616e63652053746f72616765-00000001', '/dev/nvme0n1'), 
('/dev/disk/by-id/nvme-Amazon_EC2_NVMe_Instance_Storage_AWS24D48450915FAA016', '/dev/nvme0n1'), 
('/dev/disk/by-id/nvme-nvme.1d0f-766f6c3037383162326462643766393136613265-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001', '/dev/nvme1n1'), 
('/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e', '/dev/nvme1n1'), 
('/dev/disk/by-id/nvme-nvme.1d0f-766f6c3037383162326462643766393136613265-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001-part1', '/dev/nvme1n1p1'), 
('/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e-part1', '/dev/nvme1n1p1')]
```

The 0.5 sort happens with the call to `groupby()` which then creates sublists of each of the above, based on the second element in the tuple. We get output like:

```
[('/dev/disk/by-id/nvme-nvme.1d0f-4157533234443438343530393135464141303136-416d617a6f6e20454332204e564d6520496e7374616e63652053746f72616765-00000001', '/dev/nvme0n1'), ('/dev/disk/by-id/nvme-Amazon_EC2_NVMe_Instance_Storage_AWS24D48450915FAA016', '/dev/nvme0n1')]
[('/dev/disk/by-id/nvme-nvme.1d0f-766f6c3037383162326462643766393136613265-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001', '/dev/nvme1n1'), ('/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e', '/dev/nvme1n1')]
[('/dev/disk/by-id/nvme-nvme.1d0f-766f6c3037383162326462643766393136613265-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001-part1', '/dev/nvme1n1p1'), ('/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e-part1', '/dev/nvme1n1p1')]
```

The last is `device, plain = sorted(list(group))[0]` which sorts those sublists based on the first element in the tuple, which is the device. We then select the first element in the list as to check against. This is the uniqueness checks that we used to do with /usr/bin/sort. We sort it to select devices such as `/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0781b2dbd7f916a2e` over `/dev/disk/by-id/nvme-nvme.1d0f-766f6c3037383162326462643766393136613265-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001` as they both point to '/dev/nvme1n1'.
This is how the grub2 postins.in script does it, so I am keeping it that way.

OddBloke replied:
I still don't fully understand why the order of the devices matters, but I fully appreciate the desire to stay in step with the postinst. Could we include some of this reasoning inline as comments, so folks in the future aren't similarly confused? :-)

xnox replied:
As per https://wiki.ubuntu.com/FSTAB spec, by-uuid symlink names should be preferred over by-id ones.

I wonder if our grub postinst implementation does not follow that.

matthewruffell replied:
Most devices in cloud environments won't expose by-uuid symlink names, and I do think we should keep in line with the grub2 postinst script, so we will prefer by-id names. In the latest iteration I sort the by-id entires to match the grub2 postinst script, and to keep the resulting outputs deterministic.

matthewruffell commented:
Again, thanks for the review and feedback @OddBloke, these iterations really do help improve code quality.

Changelog:
* Removed subprocess calls to sort and awk, and implemented their string manipulation in pure Python
* Dropped status_cb from the subp() call for grub-probe
* We no longer have to resolve symlinks twice.

Tested on AWS c5d.large (nitro) and t2.micro (xen) instances, LXD (with and without grub-common installed), KVM (BIOS and EFI).

How to test:
1) Reset value of grub-pc/install_devices
`$ echo reset grub-pc/install_devices | sudo debconf-communicate grub-pc`
2) Run cloud-init:
`$ sudo cloud-init single --name grub-dpkg --frequency=always`
3) Fetch value of grub-pc/install_devices
`$ echo get grub-pc/install_devices | sudo debconf-communicate grub-pc`

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
+        with suppress(FileNotFoundError):
+            for device in os.listdir("/dev/disk/by-id/"):
+                device = "{}{}".format("/dev/disk/by-id/", device)
+                devlist.append((device, str(Path(device).resolve())))
```
Following on from the above, in pre-3.6 Pythons, [`.resolve()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve) is strict and so will raise `FileNotFoundError` if the target path doesn't exist. Leaving the `suppress` in place would mean that we would ignore that error and proceed with an incomplete devlist.

matthewruffell replied:
The latest revision has changed to udevadm and we no longer use `.resolve()` or `suppress` at all now, and their imports have been removed.

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
+    # get the root disk where the /boot directory resides.
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
```
This `try` block is too large; we should have targeted exception handlers for where we specifically expect to see failures.

The contents of this try block are also a good candidate for splitting out to a separate function, I think.

matthewruffell replied:
I split the contents of the try block into a separate function, `fetch_idevs()`. I also managed to remove some duplicated code when setting `idevs_empty`.
The try block is needed as calls to `util.subp()` can throw various exceptions, namely `ProcessExecutionError`.

OddBloke commented:
Oh, and as an aside: add your GH username to `tools/.github-cla-signers` and the CLA check will stop failing.

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
+    try:
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+        # if /dev/disk/by-id/ exists, attempt to determine robust name
```
If you have the disk path, instead of guessing at what symlnks might be present, you can query udev for devlinks that are present:

```
% udevadm info --query=symlink /dev/nvme0n1
disk/by-id/nvme-SAMSUNG_MZVLB1T0HALR-000L7_S3TPNY0JB00151 disk/by-path/pci-0000:05:00.0-nvme-1 disk/by-uuid/3792634682180574674 disk/by-id/nvme-eui.0025388b710116a1 disk/by-label/rpool
```

matthewruffell replied:
This is absolutely the way to go. Thanks! Implemented in latest revision.

matthewruffell commented:
Again, @OddBloke thanks for the review! These iterations have been great, and I can see my code getting better, smaller and more robust with each feedback cycle.

To tell you the truth, I have never written a unit test in my life, but I followed the docs, looked at a lot of the existing examples and somehow managed to write one that works. It is very messy, and will likely need another revision. The tests are currently violating the 80 char line limits, and suggestions on how to fix that would be welcome.

Changelog:
* We no longer walk symlinks in /dev/disk/by-id, as we now parse the output of `udevadm info --query=symlinks` instead.
* Most imports have been removed, and we only rely on cloudinit.util now.
* Moved logic in the try block to a new function, `fetch_idevs()`
* Wrote a unit test that gets 100% coverage on `fetch_idevs()`.

matthewruffell commented:
Hi @OddBloke, please review my latest iteration. 

Changelog:
* I tidied up the unit test a little bit and added docstrings
* Added a testcase that covers SCSI drives, like what you would find on GCP. 
* I also made the strings in the unit test generic and not referencing any particular cloud vendor.

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -62,22 +86,8 @@ def handle(name, cfg, _cloud, log, _args):
-                if os.path.exists(dev):
-                    idevs = dev
-                    break
+    if idevs is None:
```
If `idevs` is not `None` but `idevs_empty` _is_, then we will attempt to pass `grub-pc grub-pc/install_devices_empty boolean None\n` into debconf-set-selections below (which I assume won't go well :-).

If `fetch_idevs` were to change from returning `(str, str)` to returning `(Optional[str], Optional[str])`[0], that would gives it the same return types as the `util.get_cfg_option_str(..., None)` calls above. This in turn could mean that the defaulting logic could move out of `fetch_idevs` and into this function (or perhaps a helper function).

Your refactor has simplified this function enough that it looks like it might be reasonable to write unit tests for it (which would hopefully catch/have caught the issue I spotted here). Would you be able to add some?

---

[0] If `Optional[str]` doesn't mean anything to you: it's the Python type annotation way of expressing "either a str or None", see [the `typing.Optional` docs](https://docs.python.org/3/library/typing.html#typing.Optional). Other languages call this Maybe or Option: https://en.wikipedia.org/wiki/Option_type

matthewruffell replied:
Great catch. I moved the logic for deciding `idevs_empty` out of `fetch_idevs` and back into `handle`, and it now conforms to how the previous implemented handled it. 

I have added a new unit test class which tests `handle()` as you requested, and it covers the test matrix of all idevs and idevs_empty configuration pairs.

OddBloke commented on cloudinit/config/tests/test_grub_dpkg.py:
```diff
@@ -0,0 +1,66 @@
+from cloudinit.config.cc_grub_dpkg import fetch_idevs
+
+
+class TestFetchIdevs():
```
```suggestion
class TestFetchIdevs:
```

OddBloke commented on cloudinit/config/tests/test_grub_dpkg.py:
```diff
@@ -0,0 +1,66 @@
+        ],
+    )
+    @mock.patch("cloudinit.config.cc_grub_dpkg.util.subp")
+    def test_fetch_idevs(self, m_subp, grub_output, udevadm_output,
```
This is a good test, thank you!

OddBloke commented on cloudinit/config/tests/test_grub_dpkg.py:
```diff
@@ -0,0 +1,66 @@
+            # KVM Instance
+            (
+                '/dev/vda',
+                '/dev/disk/by-path/pci-0000:00:00.0 /dev/disk/by-path/virtio-pci-0000:00:00.0',  # nopep8
```
As we're already within parentheses, you can wrap these long lines like this:

```suggestion
                '/dev/disk/by-path/pci-0000:00:00.0'
                ' /dev/disk/by-path/virtio-pci-0000:00:00.0',
```

Alternatively, it might be easier to read if you made this a tuple (and then did `" ".join(udevadm_output)` in the body of the test):

```suggestion
                ('/dev/disk/by-path/pci-0000:00:00.0',
                 '/dev/disk/by-path/virtio-pci-0000:00:00.0')
```

matthewruffell replied:
Thanks. Implemented in latest revision.

matthewruffell commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -40,13 +41,36 @@
+        # use udevadm to fetch /dev/disk/by-id symlinks for disk
+        devices = util.subp(['udevadm', 'info', '-r', '--query=symlink', disk],
+                            capture=True)[0].strip().split()
+        devices = list(filter(lambda dev: 'disk/by-id' in dev, devices))
```
Thanks. Implemented in latest revision.

matthewruffell commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -40,13 +41,36 @@
+        # select first device if there is one, else fall back to plain name
+        idevs = sorted(devices)[0] if devices else disk
+
+    except Exception:
```
I think I have implemented what you are asking for. Each subprocess call happens in its own try/except handler. The call to `grub-probe` also catches a `ProcessExecutionError` and handles two blue sky scenarios we would expect from a container: 1 - where grub is not installed, and 2 - where grub is installed, but /proc shows a different world view to /dev as devices from the container host are exposed to /proc. but not /dev (at least in LXD anyway).

If anything bad happens then we log the exception.

matthewruffell commented:
Hi @OddBloke, again thanks for the feedback. I have implemented the changes you have requested, please review this revision.

Changelog:
- Added unit test class TestHandle to test the handle() method in the various ways idevs and idevs_empty can be set.
- Added more comprehensive exception handling to separate out expected conditions from the unexpected. Adding appropriate logging.
- Updated TestFetchIdevs to reflect changes in exception handling, and we make sure to test all exception handling routines.
- Changed the filter statement to be more "Pythonic".

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+
+    try:
+        # get the root disk where the /boot directory resides.
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
```
On a GCE Debian buster image (the most convenient way I had of launch a Debian image), `grub-probe` is not in `PATH` of my normal user (because it's in `/usr/sbin`). While I suspect that Debian `cloud-init` units will have `/usr/sbin` in `PATH`, we could be safe and specify the full path here:

```suggestion
        disk = util.subp(['/usr/sbin/grub-probe', '-t', 'disk', '/boot'],
```

(We can't use the `util.which` helper to find it, because it relies on `PATH`.)

raharper replied:
Can we just confirm what root PATH is by default in Debian?  We consistently avoid including PATHS and expect it to include the common bin paths (including sbin).

OddBloke replied:
It is in the `PATH` of `root`, I suppose it's the problem of whoever's configuring a system if cloud-init isn't using root's `PATH`.

OddBloke commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if os.path.exists(disk):
```
This relies on the `os.path.exists("")` returning `False`; this seems like a reasonable assumption, but just in case there are unanticipated corner cases, we could be more explicit:

```suggestion
        if disk and os.path.exists(disk):
```

OddBloke commented on cloudinit/config/tests/test_grub_dpkg.py:
```diff
@@ -0,0 +1,182 @@
+            cfg_idevs_empty
+        ]
+        m_fetch_idevs.return_value = fetch_idevs_output
+        name = mock.Mock()
+        cfg = mock.Mock()
+        _cloud = mock.Mock()
+        log = mock.Mock(spec=Logger)
+        _args = mock.Mock()
+        handle(name, cfg, _cloud, log, _args)
```
If we don't need them for later use, we can just instantiate these `Mock`s in the function call:

```suggestion
        log = mock.Mock(spec=Logger)
        handle(mock.Mock(), mock.Mock(), mock.Mock(), log, mock.Mock())
```

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+        # grub-common may not be installed, especially on containers
+        # FileNotFoundError is a nested exception of ProcessExecutionError
+        if isinstance(e.reason, FileNotFoundError):
+            log.debug("grub-common is not installed, e.g. inside a container")
```
Let's not make packaging assumptions, this is really just:  'grub-probe' not found in $PATH
If we want to make statements about containers, it's worth checking directly
util.is_container().

OddBloke replied:
```suggestion
            log.debug("'grub-probe' not found in $PATH")
```

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+        # are not exposed to the container.
+        elif "failed to get canonical path" in e.stderr:
+            log.debug("Device mapping between /proc/self/mountinfo does not "
+                      "match /dev, e.g. inside a container")
```
This code does not attempt to compare /proc/self/mountinfo and host info; so I would not include this debug message; there may be other reasons for this error.

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+                                '--query=symlink', disk],
+                                capture=True)[0].strip().split()
+    except Exception:
+        util.logexc(log, "udevadm failed to gather devices for grub-dpkg")
```
udevadm DEVLINKS symlink query failed for disk=%s

OddBloke replied:
```suggestion
        util.logexc(log, "udevadm DEVLINKS symlink query failed for disk='%s'", disk)
```

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+    # filter symlinks for /dev/disk/by-id entries
+    devices = [dev for dev in devices if 'disk/by-id' in dev]
+    # select first device if there is one, else fall back to plain name
+    idevs = sorted(devices)[0] if devices else disk
```
I suspect we should log the output of udevadm info query, and our selection, otherwise we may not understand
why sometimes we have a by-id symlink, why it's different from run to run or when we have a plain disk path.

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,58 @@
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if os.path.exists(disk):
+            devices = util.subp(['udevadm', 'info', '-r',
```
use '--root' instead of -r

OddBloke replied:
```suggestion
            devices = util.subp(['udevadm', 'info', '--root',
```

raharper commented on cloudinit/config/cc_grub_dpkg.py:
```diff
@@ -43,10 +44,62 @@
+
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if disk and os.path.exists(disk):
```
If grub-probe -t returns OK, but does *not* provide a disk value (which this check seems to think is possible), should we exit earlier and just move this check up after the try/except for grub-probe?

OddBloke replied:
If `grub-probe` fails, we'd have `disk = ""` still from line 57. So it at least protects us from that. (I'm not 100% sure I've understood the request, so if this reply doesn't make sense, please do LMK.)

raharper replied:
I'm suggesting that we do this after the grub-probe call and before we try udev call.

if not disk:
   return []

This handles either scenario I was thinking of:  1) grub-probe fails, we have disks = ""  2) grub-probe returns 0, but produces no standard-output, so disk = ""

If we don't have a disk value, we might as well just return []

OddBloke replied:
Aha, right, I see what you mean. Good point, on it.
